Tdermendzhiev/ts declarations

tdermendjiev opened the pull request:
* Automate manual changes necessary for avoiding name collisions with Angular 
* Add ExtVector types to metadata and ts declarations
* Workaround missing "bool" definition

tdermendjiev committed:
WIP: Add underscore to UIEvent and AudioBuffer class names

tdermendjiev committed:
Add command line argument instead of APPLY_MANUAL_CHANGES macro

tdermendjiev committed:
Remove misleading hasNameForLinkage() check and add "bool" to typedefs

* hasNameForLinkage() sometimes returns false even if there is a name property

tdermendjiev committed:
Add a comment for the typedef _Bool workaround

mbektchiev commented on src/TypeScript/DefinitionWriter.cpp:
```diff
@@ -855,7 +942,16 @@ std::string DefinitionWriter::computeMethodReturnType(const Type* retType, const
+            
+            std::string ownerJsName = owner->jsName;
+    
+    if (apply_manual_changes) {
```
Wrong indentation.

mbektchiev commented on src/TypeScript/DefinitionWriter.cpp:
```diff
@@ -5,6 +5,8 @@
 #include <clang/AST/DeclObjC.h>
 #include <iterator>
 
+bool apply_manual_changes = false;
```
You'd better make this variable a public static field of the `DefinitionWriter` class instead of global and rename it to conform to other variables' camelCase convention.

mbektchiev commented on src/TypeScript/DefinitionWriter.cpp:
```diff
@@ -120,9 +122,20 @@ void DefinitionWriter::visit(InterfaceMeta* meta)
+        if (metaJsName == "UIEvent") {
+            metaJsName = "_UIEvent";
+        } else if (metaJsName == "HMMutableCharacteristicEvent") {
+            parametersString = "<TriggerValueType extends NSObject>";
```
Can you add a comment explaining why we have to make this change?

mbektchiev commented on src/TypeScript/DefinitionWriter.cpp:
```diff
@@ -480,10 +502,21 @@ std::string DefinitionWriter::writeMethod(MethodMeta* meta, BaseClassMeta* owner
+        }
+        
+        if (meta->jsName == "sharedKeySetForKeys") {
+            output << "<KeyType>";
```
Can you add a comment explaining why we have to make this change?

mbektchiev commented on src/TypeScript/DefinitionWriter.cpp:
```diff
@@ -480,10 +502,21 @@ std::string DefinitionWriter::writeMethod(MethodMeta* meta, BaseClassMeta* owner
+
+    if (apply_manual_changes) {
+        if (owner->jsName == "HMMutableCharacteristicEvent" && (meta->jsName == "alloc" || meta->jsName == "new")) {
+            skipGenerics = true;
```
Can you add a comment explaining why we have to make this change?

mbektchiev commented on src/TypeScript/DefinitionWriter.cpp:
```diff
@@ -496,13 +529,30 @@ std::string DefinitionWriter::writeMethod(MethodMeta* meta, BaseClassMeta* owner
     size_t lastParamIndex = meta->getFlags(::Meta::MetaFlags::MethodHasErrorOutParameter) ? (meta->signature.size() - 1) : meta->signature.size();
+    
+    if (apply_manual_changes) {
+        if (owner->jsName == "PDFView" && meta->jsName == "copy") {
```
Can you add a comment explaining why we have to make this change?

tdermendjiev committed:
Add comments for the non obvious `manual` changes

tdermendjiev committed:
Add closed generics to method names

tdermendjiev committed:
Update NSArray parameters for constructor methods

mbektchiev committed:
Fix template methods generation

* Change `typeArguments` vector's element type to `Type`. The cast in `TypeFactory::createFromObjCObjectPointerType` was incorrect in many cases.
* Add `NameRetrieverVisitor` class for getting the TypeScript or ObjectiveC name of a given type. For now it's used only in the TypeScript `DefinitionWriter`
* Make the `Type::visit` method const as it doesn't need a modifiable `this`
* Change `DefinitionWriter` to get the names of generic params using `NameRetrieverVisitor`
* Fix `getClosedGenericsIfAny` helper to support recursive generics (e.g. NSArray<NSArray<T>>)
* Fix `DefinitionWriter::writeMethod` to decrement the loop control variable after removing an item from `paramsGenerics`
* Change OSX Deployment Target Version to 10.12 in order to match LLVM's one and fix warnings

mbektchiev committed:
Change apply-manual-changes option to no-apply-manual-changes

It's better to always apply these changes and match the typings generated and
distributed in tns-typings-declaration. Still we may need to disable them for some
corner cases in the future and that's why it's good to have an opt-out mechanism.
